src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_sdk/generators/v1beta1_server.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    marker::PhantomData,
+    time::Duration,
+};
+
+use futures::{
+    channel::oneshot::{
+        self,
+        Receiver,
+        Sender,
+    },
+    Future,
+    FutureExt,
+};
+use thiserror::Error;
+use tokio::net::TcpListener;
+use tokio_stream::wrappers::TcpListenerStream;
+use tonic::{
+    transport::{
+        NamedService,
+        Server,
+    },
+    Request,
+    Response,
+};
+
+use crate::{
+    execute_rpc,
+    graplinc::grapl::api::plugin_sdk::generators::v1beta1 as native,
+    protobufs::graplinc::grapl::api::plugin_sdk::generators::v1beta1::{
+        self as proto,
+        generator_service_server::{
+            GeneratorService,
+            GeneratorServiceServer as GeneratorServiceProto,
+        },
+    },
+    protocol::{
+        healthcheck::{
+            server::init_health_service,
+            HealthcheckError,
+            HealthcheckStatus,
+        },
+        status::Status,
+    },
+    server_internals::GrpcApi,
+    SerDeError,
+};
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+pub enum GeneratorApiError {
+    #[error("failed to serialize/deserialize {0}")]
+    SerDeError(#[from] SerDeError),
+
+    #[error("received unfavorable gRPC status {0}")]
+    GrpcStatus(#[from] tonic::Status),
+}
+
+/// Implement this trait to define the API business logic
+#[tonic::async_trait]
+pub trait GeneratorApi {
+    type Error: Into<Status>;
+
+    async fn run_generator(
+        &self,
+        request: native::RunGeneratorRequest,
+    ) -> Result<native::RunGeneratorResponse, Self::Error>;
+}
+
+#[tonic::async_trait]
+impl<T> GeneratorService for GrpcApi<T>
+where
+    T: GeneratorApi + Send + Sync + 'static,
+{
+    async fn run_generator(
+        &self,
+        request: Request<proto::RunGeneratorRequest>,
+    ) -> Result<Response<proto::RunGeneratorResponse>, tonic::Status> {
+        execute_rpc!(self, request, run_generator)
+    }
+}
+
+/**
+ * !!!!! IMPORTANT !!!!!
+ * This is almost entirely cargo-culted from PipelineIngressServer.
+ * Lots of opportunities to deduplicate and simplify.
```

Comment:
Are there still opportunities to deduplicate and simplify? I think it would be a good thing to write those up as github issues. It's not clear to me what future changes this comment suggests.

Reply:
I mean, I think we're one or two things away from being able to stop copy-pasting the <T,H,F> stuff - like we are on the verge of writing that once and injecting any Api trait into it - but my rust type knowledge just isn't quite there.

Reply:
In fact, it looks like we may no longer need the PhantomData<F>?

Reply:
but yeah i'll file an issue